File: app/src/main/java/com/mee/main/videos/VideosAdapter.java
package com.mee.main.videos;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.jiajunhui.xapp.medialoader.bean.VideoItem;
import com.mee.player.databinding.VideoItemBinding;

import org.jetbrains.annotations.NotNull;

public class VideosAdapter extends ListAdapter<VideoItem, VideosAdapter.VideoItemViewHolder> {

    protected VideosAdapter(@NonNull @NotNull DiffUtil.ItemCallback<VideoItem> diffCallback) {
        super( diffCallback );
    }

    @NonNull
    @NotNull
    @Override
    public VideoItemViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent , int viewType) {
        return new VideoItemViewHolder(VideoItemBinding.inflate( LayoutInflater.from( parent.getContext() ) ));
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull VideoItemViewHolder holder , int position) {
        holder.binding.setVideoItem( getItem( position ) );
    }

    class VideoItemViewHolder extends RecyclerView.ViewHolder {
        VideoItemBinding binding;
        public VideoItemViewHolder(VideoItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

class DiffCallback extends DiffUtil.ItemCallback<VideoItem> {

    @Override
    public boolean areItemsTheSame(@NonNull @NotNull VideoItem oldItem , @NonNull @NotNull VideoItem newItem) {
        return oldItem.getPath().equals( newItem.getPath() );
    }

    @SuppressLint("DiffUtilEquals")
    @Override
    public boolean areContentsTheSame(@NonNull @NotNull VideoItem oldItem , @NonNull @NotNull VideoItem newItem) {
        return oldItem == newItem;
    }
}

File: app/src/main/java/com/mee/main/videos/VideosAdapter.kt
package com.mee.main.videos

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.net.toUri
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.mee.main.videos.VideosAdapter.VideoItemViewHolder
import com.mee.player.databinding.VideoItemBinding
import com.CodeBoy.MediaFacer.mediaHolders.videoContent
import com.mee.main.bindImage
import com.mee.main.bindVideoNameTextView
import com.mee.main.bindVideoDurationTextView

class VideosAdapter(
    val videoItemClickListener: OnClickListener,
    val moreImageViewClickListener: OnClickListener
) : ListAdapter<videoContent, VideoItemViewHolder>(DiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VideoItemViewHolder {
        return VideoItemViewHolder(VideoItemBinding.inflate(LayoutInflater.from(parent.context)))
    }

    override fun onBindViewHolder(holder: VideoItemViewHolder, position: Int) {

        bindVideoNameTextView(holder.binding.videoTitleTextView, getItem(position).videoName)
        bindImage(holder.binding.videoThumbnailImageView, getItem(position).assetFileStringUri.toUri())
        bindVideoDurationTextView(holder.binding.videoDurationTextView, getItem(position).videoDuration)


        holder.binding.moreMenuItemImageView.setOnClickListener {moreImageViewClickListener.OnClick(position)}
        holder.binding.videoItemRelativeLayout.setOnClickListener {videoItemClickListener.OnClick(position)}
    }

    class VideoItemViewHolder(var binding: VideoItemBinding) : RecyclerView.ViewHolder(binding.root)

    class DiffCallback : DiffUtil.ItemCallback<videoContent>() {
        override fun areItemsTheSame(oldItem: videoContent, newItem: videoContent): Boolean {
            return oldItem.videoId == newItem.videoId
        }

        @SuppressLint("DiffUtilEquals")
        override fun areContentsTheSame(oldItem: videoContent, newItem: videoContent): Boolean {
            return oldItem == newItem
        }
    }

    class OnClickListener(val clickListener: (position: Int) -> Unit) {
        fun OnClick(position: Int) = clickListener(position)
    }
}

